gdk/commands/component/PublishCommand.py

```diff
@@ -136,27 +136,43 @@ def _publish_component_version(self, component_name, component_version):
         PublishRecipeTransformer(self.project_config).transform()
 
         logging.info(f"Creating a new greengrass component {component_name}-{component_version}")
-        self.greengrass_client.create_gg_component()
+        self.greengrass_client.create_gg_component(self.project_config["publish_recipe_file"])
 
     def upload_artifacts_s3(self) -> None:
         """
         Uploads all the artifacts from component artifacts build folder to s3 bucket.
 
         Raises an exception when the request is not successful.
         """
-        bucket = self.project_config["bucket"]
-        region = self.project_config["region"]
+        _bucket = self.project_config["bucket"]
+
+        build_component_artifacts = list(self.project_config["gg_build_component_artifacts_dir"].iterdir())
+
+        if not build_component_artifacts:
+            logging.info("No artifacts found in the component build folder. Skipping the artifact upload step")
+            return
+
         logging.info(
-            f"Uploading component artifacts to S3 bucket: {bucket}. If this is your first time using this bucket, add the"
-            " 's3:GetObject' permission to each core device's token exchange role to allow it to download the component"
-            f" artifacts. For more information, see {utils.doc_link_device_role}."
+            (
+                "Uploading component artifacts to S3 bucket: %s. If this is your first time using this bucket, add the"
+                " 's3:GetObject' permission to each core device's token exchange role to allow it to download the component"
+                " artifacts. For more information, see %s."
+            ),
+            _bucket,
+            utils.doc_link_device_role,
         )
 
-        build_component_artifacts = list(self.project_config["gg_build_component_artifacts_dir"].iterdir())
+        self.s3_client.create_bucket(_bucket)
+
+        component_name = self.project_config["component_name"]
+        component_version = self.project_config["component_version"]
+        options = self.project_config["options"]
+        s3_upload_file_args = options.get("file_upload_args", dict())
 
-        if len(build_component_artifacts) != 0:
-            self.s3_client.create_bucket(bucket, region)
-            self.s3_client.upload_artifacts(build_component_artifacts)
+        for artifact in build_component_artifacts:
+            s3_file_path = f"{component_name}/{component_version}/{artifact.name}"
```

Comment:
should we be validating the `s3_file_path` is correctly formed e.g. customer does not have an artifact name of `my_component/`

Reply:
Here, GDK constructs the s3 file path only when the artifact exists. All the config values provided by the customer are validated at the very beginning of the command execution and component arn is constructed only after that.